components/console_cmd_mqtt/console_mqtt.c

```diff
@@ -0,0 +1,445 @@
+ * @param event_id The id for the received event.
+ * @param event_data The data for the event, esp_mqtt_event_handle_t.
+ */
+static void mqtt_event_handler(void *handler_args, esp_event_base_t base, int32_t event_id, void *event_data)
+{
+    ESP_LOGD(TAG, "Event dispatched from event loop base=%s, event_id=%" PRIi32, base, event_id);
+    esp_mqtt_event_handle_t event = event_data;
+    switch ((esp_mqtt_event_id_t)event_id) {
+    case MQTT_EVENT_BEFORE_CONNECT:
+        ESP_LOGI(TAG, "MQTT_EVENT_BEFORE_CONNECT");
+        break;
+    case MQTT_EVENT_CONNECTED:
+        ESP_LOGI(TAG, "MQTT_EVENT_CONNECTED");
+        break;
+    case MQTT_EVENT_DISCONNECTED:
+        ESP_LOGI(TAG, "MQTT_EVENT_DISCONNECTED");
+        break;
+    case MQTT_EVENT_SUBSCRIBED:
+        ESP_LOGI(TAG, "MQTT_EVENT_SUBSCRIBED, msg_id=%d", event->msg_id);
+        break;
+    case MQTT_EVENT_UNSUBSCRIBED:
+        ESP_LOGI(TAG, "MQTT_EVENT_UNSUBSCRIBED, msg_id=%d", event->msg_id);
+        break;
+    case MQTT_EVENT_PUBLISHED:
+        ESP_LOGI(TAG, "MQTT_EVENT_PUBLISHED, msg_id=%d", event->msg_id);
+        break;
+    case MQTT_EVENT_DATA:
+        ESP_LOGI(TAG, "MQTT_EVENT_DATA");
+        ESP_LOGI(TAG, "TOPIC=%.*s\r\n", event->topic_len, event->topic);
+        ESP_LOGI(TAG, "DATA=%.*s\r\n", event->data_len, event->data);
+        break;
+    case MQTT_EVENT_ERROR:
+        ESP_LOGI(TAG, "MQTT_EVENT_ERROR");
+        if (event->error_handle->error_type == MQTT_ERROR_TYPE_TCP_TRANSPORT) {
+            log_error_if_nonzero("reported from esp-tls", event->error_handle->esp_tls_last_esp_err);
+            log_error_if_nonzero("reported from tls stack", event->error_handle->esp_tls_stack_err);
+            log_error_if_nonzero("captured as transport's socket errno",  event->error_handle->esp_transport_sock_errno);
+            ESP_LOGI(TAG, "Last errno string (%s)", strerror(event->error_handle->esp_transport_sock_errno));
+        }
+        break;
+    default:
+        ESP_LOGI(TAG, "Other event id:%d", event->event_id);
+        break;
+    }
+}
+
+
+static esp_mqtt_client_config_t
+get_mqtt_config_basic(const char *uri_i)
+{
+    char *uri;
+    asprintf(&uri, "%s", uri_i);
+
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.uri = uri,
+    };
+
+    return mqtt_cfg;
+}
+
+
+static esp_mqtt_client_config_t
+get_mqtt_config_username_password(const char *uri_i, const char *username_i, const char *password_i)
+{
+    char *uri, *username, *password;
+
+    asprintf(&uri, "%s", uri_i);
+    asprintf(&username, "%s", username_i);
+    asprintf(&password, "%s", password_i);
+
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.uri = uri,
+        .credentials.username = username,
+        .credentials.authentication.password = password
+    };
+
+    return mqtt_cfg;
+}
+
+
+static void mqtt_config_free(esp_mqtt_client_config_t mqtt_cfg)
+{
+    if (mqtt_cfg.broker.address.uri) {
+        free((char *)mqtt_cfg.broker.address.uri);
```

Comment:
These structure members are just pointers to memory which this particular application didn't allocate. Therefore, they should be released by layer which allocated them. In addition, it seems that URI can be `CONFIG_MQTT_BROKER_URL` which I presume is not dynamically allocated.

Reply:
`mqtt_config_free()` is like a wrapper function. It's called from the same layer that allocated the config, i.e, `do_mqtt_cmd()`.
Since the URI can be passed from command line, I have modified the code to make all URI dynamically allocated, even CONFIG_MQTT_BROKER_URL.

Reply:
Modified the code to eleminate dynamic allocation.